crypto/manager_kmip.go

```diff
@@ -0,0 +1,423 @@
+	}
+	return &m
+}
+
+func (m ManagerKmip) ListKmipServers(
+	ctx context.Context,
+	limit *int32) ([]types.KmipClusterInfo, error) {
+
+	req := types.ListKmipServers{
+		This:  m.Reference(),
+		Limit: limit,
+	}
+	res, err := methods.ListKmipServers(ctx, m.Client(), &req)
+	if err != nil {
+		return nil, err
+	}
+	return res.Returnval, nil
+}
+
+func (m ManagerKmip) IsDefaultProviderNative(
+	ctx context.Context,
+	entity *types.ManagedObjectReference,
+	defaultsToParent bool) (bool, error) {
+
+	defaultProviderID, err := m.GetDefaultKmsClusterID(
+		ctx, entity, defaultsToParent)
+	if err != nil {
+		return false, err
+	}
+	if defaultProviderID == "" {
+		return false, nil
+	}
+	return m.IsNativeProvider(ctx, defaultProviderID)
+}
+
+func (m ManagerKmip) IsNativeProvider(
+	ctx context.Context,
+	providerID string) (bool, error) {
+
+	info, err := m.GetClusterStatus(ctx, providerID)
+	if err != nil {
+		return false, err
+	}
+	if info == nil {
+		return false, nil
+	}
+	return info.ManagementType == string(
+		types.KmipClusterInfoKmsManagementTypeNativeProvider), nil
+}
+
+func (m ManagerKmip) GetDefaultKmsClusterID(
+	ctx context.Context,
+	entity *types.ManagedObjectReference,
+	defaultsToParent bool) (string, error) {
+
+	req := types.GetDefaultKmsCluster{
+		This:             m.Reference(),
+		Entity:           entity,
+		DefaultsToParent: &defaultsToParent,
+	}
+	res, err := methods.GetDefaultKmsCluster(ctx, m.Client(), &req)
+	if err != nil {
+		return "", err
+	}
+	if res.Returnval != nil {
+		return res.Returnval.Id, nil
+	}
+	return "", nil
+}
+
+func (m ManagerKmip) GetStatus(
+	ctx context.Context,
+	clusters ...types.KmipClusterInfo) ([]types.CryptoManagerKmipClusterStatus, error) {
+
+	req := types.RetrieveKmipServersStatus_Task{
+		This:     m.Reference(),
+		Clusters: clusters,
+	}
+	res, err := methods.RetrieveKmipServersStatus_Task(ctx, m.Client(), &req)
+	if err != nil {
+		return nil, err
+	}
+
+	task := object.NewTask(m.Client(), res.Returnval)
+	taskInfo, err := task.WaitForResult(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	if taskInfo.Result == nil {
+		return nil, nil
+	}
+	result, ok := taskInfo.Result.(types.ArrayOfCryptoManagerKmipClusterStatus)
+	if !ok {
+		return nil, nil
+	}
+	if len(result.CryptoManagerKmipClusterStatus) == 0 {
+		return nil, nil
+	}
+
+	return result.CryptoManagerKmipClusterStatus, nil
+}
+
+func (m ManagerKmip) GetClusterStatus(
+	ctx context.Context,
+	providerID string) (*types.CryptoManagerKmipClusterStatus, error) {
+
+	result, err := m.GetStatus(
+		ctx,
+		types.KmipClusterInfo{
+			ClusterId: types.KeyProviderId{
+				Id: providerID,
+			},
+		})
+	if err != nil {
+		return nil, err
+	}
+	if len(result) == 0 {
+		return nil, fmt.Errorf("invalid cluster ID")
+	}
+	return &result[0], nil
+}
+
+func (m ManagerKmip) GetServerStatus(
+	ctx context.Context,
+	providerID, serverName string) (*types.CryptoManagerKmipServerStatus, error) {
```

Comment:
Don't see where `serverName` is used here, looks like it always returns the 1st server?